include/lbann/utils/gpu/rocm.hpp

```diff
@@ -0,0 +1,212 @@
+                  + std::string(")"));                          \
+    }                                                           \
+    LBANN_ROCM_SYNC(false);                                     \
+  } while (0)
+#define FORCE_CHECK_ROCM_NOSYNC(rocm_call)                      \
+  do {                                                          \
+    hipError_t status_CHECK_ROCM = (rocm_call);                 \
+    if (status_CHECK_ROCM != hipSuccess) {                      \
+      LBANN_ERROR(std::string("ROCm error (")                   \
+                  + hipGetErrorString(status_CHECK_ROCM)        \
+                  + std::string(")"));                          \
+    }                                                           \
+  } while (0)
+#ifdef LBANN_DEBUG
+#define CHECK_ROCM(rocm_call) FORCE_CHECK_ROCM(rocm_call);
+#else
+#define CHECK_ROCM(rocm_call) FORCE_CHECK_ROCM_NOSYNC(rocm_call)
+#endif // #ifdef LBANN_DEBUG
+
+namespace lbann {
+namespace rocm {
+
+// -------------------------------------------------------------
+// Wrapper classes
+// -------------------------------------------------------------
+
+/** Wrapper class for a HIP event. */
+class event_wrapper {
+public:
+  event_wrapper();
+  event_wrapper(const event_wrapper& other);
+  event_wrapper& operator=(const event_wrapper& other);
+  ~event_wrapper();
+  /** Enqueue HIP event on a HIP stream. */
+  void record(hipStream_t stream);
+  /** Check whether HIP event has completed. */
+  bool query() const;
+  /** Wait until HIP event has completed. */
+  void synchronize();
+  /** Get HIP event object. */
+  hipEvent_t& get_event();
+private:
+  /** HIP event object.
+   *  The event object lifetime is managed internally.
+   */
+  hipEvent_t m_event;
+  /** HIP stream object.
+   *  The stream object lifetime is assumed to be managed externally.
+   */
+  hipStream_t m_stream;
+};
+
+// -------------------------------------------------------------
+// Helper functions for tensor operations
+// -------------------------------------------------------------
+
+/** Copy entries between GPU tensors. */
+template <typename TensorDataType>
+void copy_tensor(
+  hipStream_t stream,
+  const std::vector<size_t>& dims,
+  const TensorDataType* input,
+  const std::vector<size_t>& input_strides,
+  TensorDataType* output,
+  const std::vector<size_t>& output_strides);
+
+// -------------------------------------------------------------
+// Utilities for Thrust
+// -------------------------------------------------------------
+namespace thrust {
+
+/** Thrust execution policy. */
+using execute_on_stream
+#if THRUST_MAJOR_VERSION > 1 || THRUST_MINOR_VERSION >= 9
+  = ::thrust::hip_rocprim::execute_on_stream; // >= 1.9.1
+#elif THRUST_MAJOR_VERSION == 1 && THRUST_MINOR_VERSION == 8
+  = ::thrust::system::hip::detail::execute_on_stream;
+#else
+  = std::nullptr_t;
+  static_assert(false, "Thrust 1.8 or newer is required");
+#endif
```

Comment:
Can we assume Thrust >=1.9 for ROCm >=3.8?
```suggestion
using execute_on_stream = ::thrust::hip_rocprim::execute_on_stream;
```

Reply:
I think we can, good catch!